sdk/lib/_internal/js_dev_runtime/private/js_string.dart

```diff
@@ -83,6 +83,104 @@ final class JSString extends Interceptor
     return stringReplaceAllFuncUnchecked(this, from, onMatch, onNonMatch);
   }
 
+  @notNull
+  Iterable<T> splitMap<T>(
+    Pattern pattern, {
+    T Function(Match match)? onMatch,
+    T Function(String nonMatch)? onNonMatch,
+  }) {
+    return stringSplitMapUnchecked(this, pattern, onMatch, onNonMatch);
+  }
+
+  Iterable<T> stringSplitMapUnchecked<T>(String receiver, Pattern pattern,
+      T Function(Match)? onMatch, T Function(String)? onNonMatch) {
+    onMatch ??= (match) => match[0]! as T;
+    onNonMatch ??= (string) => string as T;
+    if (pattern is String) {
+      return stringSplitStringMapUnchecked(
+        receiver, pattern, onMatch, onNonMatch);
+    }
+    if (pattern is JSSyntaxRegExp) {
+      return stringSplitJSRegExpMapUnchecked(receiver, pattern, onMatch, onNonMatch);
+    }
+    return stringSplitGeneralMapUnchecked(receiver, pattern, onMatch, onNonMatch);
+  }
+
+  Iterable<T> stringSplitStringMapUnchecked<T>(String receiver, String pattern,
+      T Function(Match) onMatch, T Function(String) onNonMatch) {
+    if (pattern.isEmpty) {
+      return stringSplitEmptyMapUnchecked(receiver, onMatch, onNonMatch);
+    }
+    List<T> result = <T>[];
+    int startIndex = 0;
+    int patternLength = pattern.length;
+    while (true) {
+      int position = stringIndexOfStringUnchecked(receiver, pattern, startIndex);
+      if (position == -1) {
+        break;
+      }
+      result.add(onNonMatch(receiver.substring(startIndex, position)));
+      result.add(onMatch(StringMatch(position, receiver, pattern)));
+      startIndex = position + patternLength;
+    }
+    result.add(onNonMatch(receiver.substring(startIndex)));
+    return result;
+  }
+
+  Iterable<T> stringSplitEmptyMapUnchecked<T>(
+      String receiver, T Function(Match) onMatch, T Function(String) onNonMatch) {
+    List<T> result = <T>[];
+    int length = receiver.length;
+    int i = 0;
+    result.add(onNonMatch(""));
+    while (i < length) {
+      result.add(onMatch(StringMatch(i, receiver, "")));
+      // Special case to avoid splitting a surrogate pair.
```

Comment:
A `string.split("")` would split surrogate pairs. So would `string.split(RegExp(r'(?:)'))` without the `unicode:true` flag.

(It'd be nice if it didn't, but then it'd be even niceer if it didn't break grapheme clusters either.)